pkg/healthchecker/health_checker.go

```diff
@@ -0,0 +1,167 @@
+limitations under the License.
+*/
+
+package healthchecker
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/node-problem-detector/cmd/healthchecker/options"
+	"k8s.io/node-problem-detector/pkg/healthchecker/types"
+)
+
+type healthChecker struct {
+	enableRepair    bool
+	healthCheckFunc func(time.Duration) bool
+	// The repair is "best-effort" and ignores the error from the underlying actions.
+	// The bash commands to kill the process will fail if the service is down and hence ignore.
+	repairFunc         func()
+	uptimeFunc         func() (time.Duration, error)
+	crictlPath         string
+	healthCheckTimeout time.Duration
+	coolDownTime       time.Duration
+}
+
+// NewHealthChecker returns a new health checker configured with the given options.
+func NewHealthChecker(hco *options.HealthCheckerOptions) (types.HealthChecker, error) {
+	hc := &healthChecker{
+		enableRepair:       hco.EnableRepair,
+		crictlPath:         hco.CriCtlPath,
+		healthCheckTimeout: hco.HealthCheckTimeout,
+		coolDownTime:       hco.CoolDownTime,
+	}
+	hc.healthCheckFunc = getHealthCheckFunc(hco)
+	hc.repairFunc = getRepairFunc(hco)
+	hc.uptimeFunc = getUptimeFunc(hco.CmdTimeout, hco.SystemService)
+	return hc, nil
+}
+
+// getUptimeFunc returns the time for which the given service has been running.
+func getUptimeFunc(timeout time.Duration, service string) func() (time.Duration, error) {
+	return func() (time.Duration, error) {
+		out, err := execCommand(timeout, "systemctl", "show", service, "--property=ActiveEnterTimestamp")
+		if err != nil {
+			return time.Duration(0), err
+		}
+		val := strings.Split(out, "=")
+		if len(val) < 2 {
+			return time.Duration(0), errors.New("could not parse the service uptime time correctly")
+		}
+		t, err := time.Parse(types.UptimeTimeLayout, val[1])
+		if err != nil {
+			return time.Duration(0), err
+		}
+		return time.Since(t), nil
+	}
+}
+
+// getRepairFunc returns the repair function based on the component.
+func getRepairFunc(hco *options.HealthCheckerOptions) func() {
+	switch hco.Component {
+	case types.DockerComponent:
+		// Use "docker ps" for docker health check. Not using crictl for docker to remove
+		// dependency on the kubelet.
+		return func() {
+			execCommand(hco.CmdTimeout, "pkill", "-SIGUSR1", "dockerd")
+			execCommand(hco.CmdTimeout, "systemctl", "kill", "--kill-who=main", hco.SystemService)
+		}
+	default:
+		// Just kill the service for all other components
+		return func() {
+			execCommand(hco.CmdTimeout, "systemctl", "kill", "--kill-who=main", hco.SystemService)
+		}
+	}
+}
+
+// getHealthCheckFunc returns the health check function based on the component.
+func getHealthCheckFunc(hco *options.HealthCheckerOptions) func(time.Duration) bool {
+	switch hco.Component {
+	case types.KubeletComponent:
+		return kubeletHealthCheck
+	case types.DockerComponent:
+		return func(timeout time.Duration) bool {
```

Comment:
1) Can we make the way we use timeout consistent among the 3 functions? Maybe outside function accept `HealthCheckerOptions`, internal function doesn't accept arguments?
2) What is the relationship between `HealthCheckTimeout` and `CmdTimeout`? It is a bit confusing. :P

Reply:
1. Done. Made the kubelet function inline.
2. HealthCheckTimeout is used to timeout actions in health check function. CmdTimeout is the timeout used for all other CLI commands run. For example : getUptime. Kept these two different because we might want to have a longer timeout on health check than other commands.

Reply:
I know what they are, but it is hard for users to understand when see the flag description.

Do we need to make command timeout configurable? Maybe have a constant value for it? I don't think people will need to tweak it. If they do, we can think about how to name that flag then.

Reply:
Sounds good. Made it a constant and removed from the flags.